File: sdks/java/src/test/java/io/github/cuihairu/croupier/sdk/FilePushTest.java
package io.github.cuihairu.croupier.sdk;

import io.github.cuihairu.croupier.sdk.transport.Protocol;
import io.github.cuihairu.croupier.sdk.wire.SdkWireMessages;
import org.junit.jupiter.api.Test;
import java.lang.reflect.Method;
import org.junit.jupiter.api.io.TempDir;

import java.nio.file.Files;
import java.nio.file.Path;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * F：文件下发接收（hotpatch P1 传输层）——Java 实现测试。
 */
public class FilePushTest {

    private CroupierClientImpl newClient(boolean enable, String staging) {
        ClientConfig config = new ClientConfig();
        config.setEnableFileTransfer(enable);
        config.setMaxFileSize(1024);
        config.setFileStagingDir(staging);
        return new CroupierClientImpl(config);
    }

    private byte[] push(CroupierClientImpl client, SdkWireMessages.FilePushRequest request)
            throws Exception {
        Method m = CroupierClientImpl.class.getDeclaredMethod("handleFilePushRequest", byte[].class);
        m.setAccessible(true);
        return (byte[]) m.invoke(client, (Object) SdkWireMessages.encodeFilePushRequest(request));
    }

    private SdkWireMessages.FilePushResponse decode(byte[] raw) {
        return SdkWireMessages.decodeFilePushResponse(raw);
    }

    private String sha256(byte[] data) throws Exception {
        byte[] digest = java.security.MessageDigest.getInstance("SHA-256").digest(data);
        StringBuilder sb = new StringBuilder(digest.length * 2);
        for (byte b : digest) {
            sb.append(String.format("%02x", b));
        }
        return sb.toString();
    }

    @Test
    public void validPayloadStagedAndOk(@TempDir Path tempDir) throws Exception {
        Path staging = tempDir.resolve("staging");
        CroupierClientImpl client = newClient(true, staging.toString());
        byte[] data = "print('hotfix')".getBytes();
        byte[] raw = push(client, new SdkWireMessages.FilePushRequest(
                "t-1", "hotfix.lua", sha256(data), data));
        SdkWireMessages.FilePushResponse response = decode(raw);
        assertTrue(response.ok);
        assertTrue(response.storedPath.endsWith("hotfix.lua"));
        assertEquals("print('hotfix')", Files.readString(Path.of(response.storedPath)));
    }

    @Test
    public void disabledFlagRejects(@TempDir Path tempDir) throws Exception {
        CroupierClientImpl client = newClient(false, tempDir.toString());
        byte[] data = "x".getBytes();
        SdkWireMessages.FilePushResponse response = decode(push(client,
                new SdkWireMessages.FilePushRequest("t-2", "a.lua", sha256(data), data)));
        assertFalse(response.ok);
        assertTrue(response.error.contains("file transfer is disabled"));
    }

    @Test
    public void checksumMismatchRejects(@TempDir Path tempDir) throws Exception {
        CroupierClientImpl client = newClient(true, tempDir.toString());
        byte[] raw = push(client, new SdkWireMessages.FilePushRequest(
                "t-3", "a.lua", "de".repeat(32), "x".getBytes()));
        SdkWireMessages.FilePushResponse response = decode(raw);
        assertFalse(response.ok);
        assertTrue(response.error.contains("checksum mismatch"));
    }

    @Test
    public void pathTraversalRejects(@TempDir Path tempDir) throws Exception {
        CroupierClientImpl client = newClient(true, tempDir.toString());
        byte[] data = "x".getBytes();
        for (String evil : new String[]{"../evil.lua", "sub/dir/evil.lua", "/etc/evil.lua"}) {
            byte[] raw = push(client, new SdkWireMessages.FilePushRequest(
                    "t-4", evil, sha256(data), data));
            SdkWireMessages.FilePushResponse response = decode(raw);
            assertFalse(response.ok);
            assertTrue(response.error.contains("bare basename"), "evil=" + evil);
        }
        assertFalse(Files.exists(tempDir.resolve("evil.lua")));
    }

    @Test
    public void oversizeRejects(@TempDir Path tempDir) throws Exception {
        CroupierClientImpl client = newClient(true, tempDir.toString());
        byte[] data = new byte[2048];
        byte[] raw = push(client, new SdkWireMessages.FilePushRequest(
                "t-5", "big.lua", sha256(data), data));
        SdkWireMessages.FilePushResponse response = decode(raw);
        assertFalse(response.ok);
        assertTrue(response.error.contains("exceeds max"));
    }

    @Test
    public void emptyTransferIdRejects(@TempDir Path tempDir) throws Exception {
        CroupierClientImpl client = newClient(true, tempDir.toString());
        byte[] data = "x".getBytes();
        byte[] raw = push(client, new SdkWireMessages.FilePushRequest(
                "", "a.lua", sha256(data), data));
        SdkWireMessages.FilePushResponse response = decode(raw);
        assertFalse(response.ok);
        assertTrue(response.error.contains("transferId is required"));
    }
}

File: sdks/go/pkg/croupier/file_push_test.go
package croupier

import (
	"crypto/sha256"
	"encoding/hex"
	"os"
	"path/filepath"
	"strings"
	"testing"

	"github.com/cuihairu/croupier/sdks/go/pkg/croupier/protocol"
	sdkv1 "github.com/cuihairu/croupier/sdks/go/pkg/pb/croupier/sdk/v1"
)

func sha256Hex(data []byte) string {
	sum := sha256.Sum256(data)
	return hex.EncodeToString(sum[:])
}

func newFilePushManager(t *testing.T, enable bool) *TCPManager {
	t.Helper()
	config := ClientConfig{
		EnableFileTransfer: enable,
		MaxFileSize:        1024,
		FileStagingDir:     filepath.Join(t.TempDir(), "staging"),
	}
	manager, err := NewTCPManager(config, map[string]FunctionHandler{})
	if err != nil {
		t.Fatalf("NewTCPManager: %v", err)
	}
	m := manager.(*TCPManager)
	m.functions = []*sdkv1.ProviderFunctionDescriptor{{Id: "f1"}}
	return m
}

// F：文件下发原语——编解码回路。
func TestFilePushCodecRoundTrip(t *testing.T) {
	req := &filePushRequest{
		transferID:    "t-1",
		fileName:      "patch.lua",
		contentSha256: "abc123",
		data:          []byte("payload-bytes"),
	}
	decoded, err := decodeFilePushRequest(encodeFilePushRequest(req))
	if err != nil {
		t.Fatalf("decode: %v", err)
	}
	if decoded.transferID != "t-1" || decoded.fileName != "patch.lua" ||
		decoded.contentSha256 != "abc123" || string(decoded.data) != "payload-bytes" {
		t.Fatalf("round trip mismatch: %+v", decoded)
	}
}

func TestFilePushResponseCodec(t *testing.T) {
	resp := encodeFilePushResponse(filePushResponse{
		transferID: "t-9",
		ok:         true,
		storedPath: "/staging/patch.lua",
	})
	if resp[0] != 0x0A {
		t.Fatalf("field1 tag = %x", resp[0])
	}
	body := string(resp)
	if !strings.Contains(body, "t-9") || !strings.Contains(body, "/staging/patch.lua") {
		t.Fatalf("response content mismatch: %x", resp)
	}
	if !strings.Contains(body, string([]byte{0x10, 0x01})) {
		t.Fatalf("ok=true not encoded: %x", resp)
	}
}

// F：安全约束——每条一条用例。
func TestFilePushValidation(t *testing.T) {
	validFile := []byte("print('hotfix')")
	validSha := sha256Hex(validFile)

	cases := []struct {
		name       string
		enable     bool
		transferID string
		fileName   string
		sha        string
		data       []byte
		errSubstr  string
	}{
		{"disabled flag", false, "t-1", "patch.lua", validSha, validFile, "file transfer is disabled"},
		{"empty transfer_id", true, "", "patch.lua", validSha, validFile, "transfer_id is required"},
		{"path traversal ../", true, "t-1", "../evil.lua", validSha, validFile, "bare basename"},
		{"absolute path", true, "t-1", "/etc/evil.lua", validSha, validFile, "bare basename"},
		{"subdir escape", true, "t-1", "sub/dir/evil.lua", validSha, validFile, "bare basename"},
		{"dotdot name rejected", true, "t-1", "..evil", validSha, validFile, "bare basename"},
		{"empty payload", true, "t-1", "patch.lua", validSha, nil, "file payload is empty"},
		{"oversize", true, "t-1", "patch.lua", validSha, make([]byte, 2048), "exceeds max"},
		{"missing sha", true, "t-1", "patch.lua", "", validFile, "content_sha256 is required"},
		{"checksum mismatch", true, "t-1", "patch.lua", strings.Repeat("ab", 32), validFile, "checksum mismatch"},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			m := newFilePushManager(t, tc.enable)
			req := &filePushRequest{
				transferID:    tc.transferID,
				fileName:      tc.fileName,
				contentSha256: tc.sha,
				data:          tc.data,
			}
			err := m.validateFilePush(req)
			if tc.errSubstr == "" {
				if err != nil {
					t.Fatalf("expected pass, got %v", err)
				}
				return
			}
			if err == nil || !strings.Contains(err.Error(), tc.errSubstr) {
				t.Fatalf("expected error containing %q, got %v", tc.errSubstr, err)
			}
		})
	}
}

// F：落盘行为——原子写 + 暂存目录落点。
func TestFilePushWritesToStaging(t *testing.T) {
	m := newFilePushManager(t, true)
	fileName := "hotfix.lua"
	data := []byte("print('hotfix')")
	body := encodeFilePushRequest(&filePushRequest{
		transferID:    "t-7",
		fileName:      fileName,
		contentSha256: sha256Hex(data),
		data:          data,
	})
	respBody, err := m.handleFilePushRequest(body)
	if err != nil {
		t.Fatalf("handleFilePushRequest: %v", err)
	}
	// 确认帧应包含 ok=true 与暂存路径
	resp := string(respBody)
	if !strings.Contains(resp, "hotfix.lua") {
		t.Fatalf("stored path missing in response: %x", respBody)
	}
	stored, readErr := os.ReadFile(filepath.Join(m.fileStagingDir(), fileName))
	if readErr != nil {
		t.Fatalf("staged file not found: %v", readErr)
	}
	if string(stored) != string(data) {
		t.Fatalf("staged content mismatch: %q", stored)
	}
}

// F：agent 帧回路——FilePushRequest 帧进、FilePushResponse 帧出。
func TestFilePushFrameRoundTrip(t *testing.T) {
	m := newFilePushManager(t, true)
	data := []byte("code-v2")
	body := encodeFilePushRequest(&filePushRequest{
		transferID:    "t-frame",
		fileName:      "code.lua",
		contentSha256: sha256Hex(data),
		data:          data,
	})
	wire := protocol.NewMessageBody(protocol.MsgProviderFilePushRequest, 42, body)

	_, msgID, reqID, payload, err := protocol.ParseMessageFromBody(wire)
	if err != nil {
		t.Fatalf("parse frame: %v", err)
	}
	if msgID != protocol.MsgProviderFilePushRequest {
		t.Fatalf("msgID = %x", msgID)
	}
	respBody, err := m.handleFilePushRequest(payload)
	if err != nil {
		t.Fatalf("handle: %v", err)
	}
	respFrame := protocol.NewMessageBody(protocol.GetResponseMsgID(msgID), reqID, respBody)
	_, respMsgID, respReqID, respPayload, err := protocol.ParseMessageFromBody(respFrame)
	if err != nil {
		t.Fatalf("parse response frame: %v", err)
	}
	if respMsgID != protocol.MsgProviderFilePushResponse {
		t.Fatalf("response msgID = %x", respMsgID)
	}
	if respReqID != 42 {
		t.Fatalf("reqID = %d", respReqID)
	}
	if len(respPayload) == 0 {
		t.Fatal("empty response payload")
	}
}
